profiling/src/allocation.rs

```diff
@@ -145,47 +170,50 @@ pub fn allocation_profiling_rinit() {
         return;
     }
 
-    unsafe {
-        HEAP = Some(zend::zend_mm_get_heap());
-    }
-
-    if !is_zend_mm() {
-        // Neighboring custom memory handlers found
-        debug!("Found another extension using the ZendMM custom handler hook");
+    ZEND_MM_STATE.with(|cell| {
+        let zend_mm_state = cell.get();
         unsafe {
-            zend::zend_mm_get_custom_handlers(
-                // Safety: `unwrap()` is safe here, as `HEAP` is initialized just above
-                HEAP.unwrap(),
-                &mut PREV_CUSTOM_MM_ALLOC,
-                &mut PREV_CUSTOM_MM_FREE,
-                &mut PREV_CUSTOM_MM_REALLOC,
-            );
-            ALLOCATION_PROFILING_ALLOC = allocation_profiling_prev_alloc;
-            ALLOCATION_PROFILING_FREE = allocation_profiling_prev_free;
-            ALLOCATION_PROFILING_REALLOC = allocation_profiling_prev_realloc;
-            ALLOCATION_PROFILING_PREPARE_ZEND_HEAP = prepare_zend_heap_none;
-            ALLOCATION_PROFILING_RESTORE_ZEND_HEAP = restore_zend_heap_none;
+            (*zend_mm_state).heap = Some(zend::zend_mm_get_heap());
         }
-    } else {
-        unsafe {
-            ALLOCATION_PROFILING_ALLOC = allocation_profiling_orig_alloc;
-            ALLOCATION_PROFILING_FREE = allocation_profiling_orig_free;
-            ALLOCATION_PROFILING_REALLOC = allocation_profiling_orig_realloc;
-            ALLOCATION_PROFILING_PREPARE_ZEND_HEAP = prepare_zend_heap;
-            ALLOCATION_PROFILING_RESTORE_ZEND_HEAP = restore_zend_heap;
+
+        if !is_zend_mm() {
+            // Neighboring custom memory handlers found
+            debug!("Found another extension using the ZendMM custom handler hook");
+            unsafe {
+                zend::zend_mm_get_custom_handlers(
+                    // Safety: `unwrap()` is safe here, as `heap` is initialized just above
+                    (*zend_mm_state).heap.unwrap(),
```

Comment:
Could probably just save the un-optioned heap from above and avoid the pointer deref + unwrap? Same goes for the `ddog_php_prof_zend_mm_set_custom_handlers` section below.

Reply:
Done, I also added more docs